src/identity.rs

```diff
@@ -0,0 +1,117 @@
+use linkerd2_proxy_api::identity::{
+    self as pb,
+    identity_server::{Identity, IdentityServer},
+};
+use std::{
+    collections::HashMap,
+    fs::File,
+    io::{self, BufReader, ErrorKind},
+    net::SocketAddr,
+    path::{Path, PathBuf},
+    time::{Duration, SystemTime},
+};
+use tracing_futures::Instrument;
+
+pub type Error = Box<dyn std::error::Error + Send + Sync + 'static>;
+
+pub struct IdentityService {
+    identities: HashMap<String, Certificates>,
+}
+
+struct Certificates {
+    leaf: Vec<u8>,
+    intermediates: Vec<Vec<u8>>,
+}
+
+impl IdentityService {
+    pub fn new(mut path: PathBuf) -> Result<IdentityService, io::Error> {
+        let mut identities = HashMap::new();
+
+        for entry in path.read_dir().expect("read_dir call failed") {
+            if let Ok(entry) = entry {
+                if let Ok(file_type) = entry.file_type() {
+                    if file_type.is_dir() {
+                        path.push(entry.file_name());
+                        path.push("crt.pem");
+                        let certs = Certificates::load(&path)
+                            .expect("failed to load certificates from crt.pem");
+                        let local_name = entry.file_name().into_string().unwrap();
+                        tracing::info!(?local_name, "added");
+                        identities.insert(local_name, certs);
+                    }
+                }
+            }
+        }
```

Comment:
If I'm reading this correctly, it seems like we expect an identity dir containing certs for "foo" and "bar" to be laid out like this:
```
identity/
 |-- foo/
 |    |-- crt.pem
 |-- bar/
      |-- crt.pem
```

Is there a reason we need a separate directory for each identity, if it will always only ever contain one file? Naively, it seems like it could just be
```
identity/
 |-- foo.pem
 |-- bar.pem
```

This would simplify the logic here, and also make setting up the identity dir for tests slightly simpler IMO.

If there's a reason not to do that, it would be nice to know why.

Reply:
So it's not very obvious from the CLI documentation, but it's likely that each directory corresponding to an identity will have more than one file.

This is because right now, this is coupled with the change in olix0r/l2-proxy-harness#2 where the directory given is expected to have a `csr.der` and `key.p8`.

So for mock identity to work, you are correct that only the `crt.pem` is needed.

For the change to work in proxy-harness, the directory also needs a `csr.der` and `key.p8`.

Right now I use the same directory for this, so each directory that corresponds to an identity has these three files.

Now that I'm writing this out, maybe it makes sense to keep these assumptions separate.

Reply:
Okay, yeah, in context this makes perfect sense. Disregard me!